### ISSUE TITLE
new DrawImage options API (proposal)

### ISSUE THREAD

**Issue by eihigh** [hajimehoshi/ebiten#752]
Hi, I'm developing my games on Ebiten everyday. Thank you a lot.

Now I want to propose a new DrawImage API, `DrawGroup`.

My external (and experimental) implementation is https://github.com/eihigh/love-and-hate/blob/master/internal/draw/draw.go .

## Design

`DrawGroup` contains `Dst`, `ColorM`, `CompositeMode`, and `Filter`. They're based on [Performance Tips](https://github.com/hajimehoshi/ebiten/wiki/Performance-Tips) .

```Go
func draw() {
	// If Dst is nil, screen is used as a default dst.
	dg := &ebiten.DrawGroup{Dst: nil}

	// Simple drawing.
	dg.DrawImage(someImage)

	// The first argument is *ebiten.Image.
	// The following arguments are GeoM/ColorM operations.
	dg.DrawImage(
		someImage,
		ebiten.Translate(0, 100),
		ebiten.Scale(-1, 1),
	)

	// Another group.
	// The less the group, the more efficient.
	dg = &ebiten.DrawImage{Mode: ebiten.CompositeModeLighter}

	// Each DrawImages are independent.
	// So there is no need to `op.GeoM.Reset()`
	dg.DrawImage(
		someEffect,
		ebiten.Translate(50, 50),
		ebiten.ScaleColor(1, 1, 1, 0.5),
	)
}

// Users can define custom operations.
func center(i *ebiten.Image) ebiten.OptionFn {
	w, h := i.Size()
	return ebiten.Translate(-float64(w)/2, -float64(h)/2)
}
```

By doing this,

* more obvious about drawing performance
* `screen` is hidden; never use screen as source
    * and no need to manage `screen *ebiten.Image` passed to the update function
* reduce boring `op.GeoM.Reset(); op.GeoM.Reset(); ...`
* Tell deprecated fields in DrawImageOptions goodbye.

Concerns:

* Is this "functional option pattern" really easy (especially for beginners?)
* Is this really simple?
* Concurrency (maybe Dst should not be nil, set screen)
* Some API changes, too many API additions.
* My bad English ;)

**Comment by hajimehoshi** [comment]
Hi, thank you for using Ebiten, and proposal!

My concern is that this sounds like stateful. For example, I was not sure if `ebiten.Translate(50, 50)` means that this translates from the last position or the origin position (0, 0).

Probably what you suggest is similar to Canvas's graphics context API. Correct me if I'm wrong. Yes, the current API is stateless but might be hard to follow the performance tips.

> reduce boring `op.GeoM.Reset(); op.GeoM.Reset();` ...

I think you don't have to call them so often if you re-create `DrawImageOptions`, but I think your point is to continue to use other properties like `ColorM` to follow the performance tip.

> Tell deprecated fields in DrawImageOptions goodbye.

They will be removed at Ebiten 2.0, probably next year :-)

> Is this "functional option pattern" really easy (especially for beginners?)

Well, functional option pattern is very popular among Go users. I feel like this should be a way to go, regardless of whether we adopt your suggested API or not.

> Is this really simple?

Not only the simplicity, but I'd also be concerned about stateful-ness.

> Concurrency (maybe Dst should not be nil, set screen)

I think this is acceptable as long as the API doc explains about the concurrent (non-)safety.

> My bad English ;)

No problem at all!

**Comment by hajimehoshi** [comment]
I'd like to confirm what you are trying to solve:

* Improving API usability: it makes sense if you think stateful API is better than stateless, but this sounds controversial topic. Ebiten tends to like stateless API anyway.
* Making it easy to follow the performance tips: I think the current API is problematic in terms of this, but the performance tips is an implementation detail and can be changed any time. Exposing this notion to the public API might be problematic in the future. This is also arguable.

**Comment by eihigh** [comment]
Thanks for your response and I'm sorry for I wasted your time 😢

I agree that API should be stateless. And I should re-think about it.
I have to study more and more...

I'll close this issue. Thanks!

**Comment by hajimehoshi** [comment]
I didn't mean to blame you, I just wanted to understand your suggestion and confirm if my understanding was correct.

Rather, I appreciate you because I now realized that API that reflects the performance tips might be useful.

Thank you again!
